### DISCUSSION TITLE
Interested in dvbs2x ldpc code

### DISCUSSION THREAD

**Discussion by nbecker** [NVlabs/sionna#9] [General]
Just learning about sionna, it looks amazing!

I'm interested in satellite applications, where dvbs2x is the common standard.  In order to move forward on this, we need the dvbs2x ldpc code and constellations.  Sionna has an ldpc encoder (LDPC5GEncoder) but unfortunately it is specific to 5gnr code.  This includes logic to select the parity check matrix, and to perform rate matching.

On the decoder side there is LDPCBPDecoder which looks generic and  which perhaps can be used for dvbs2x without change, just by replacing the parity check matrix.

Ideally, I think, there would be a generic ldpc encoder and ldpc decoder.  All rate matching and adaptation would then be done in separate functions.  This would make it easy to support other wireless standard ldpc codes.
There could be a generic LDPCEncoder which would be used by LDPC5GEncoder and others (such as dvbs2x), so some refactoring is needed.  On the decoder side it appears at quick glance that the refactoring is already done and the generic decoder is already implemented (although I haven't studied this in detail).

Any thoughts?

Best,
Neal

**Comment by SebastianCa** [comment]
We are happy to hear that you like Sionna. Indeed, the LDPCBPDecoder is generic and can be used with arbitrary (binary) parity-check matrices. Further, the alist import function allows to load parity-check matrices directly as “alist” file format from external sources such as [1].
 
A generic encoder is currently under active development and will be integrated in Sionna soon. The problem however is that generic encoding has a higher computational complexity than the LDPC5GEncoder which is carefully optimized for the underlying code structure (so-called “Richardson-Urbanke” encoding see [2]). Please keep in mind that for many research applications the “all-zero codeword trick” allows to simulate BER/BLER curves without the need of an actual encoder. You can find an example in the BICM example notebook [3].
 
[1] https://www.uni-kl.de/channel-codes/ml-simulation-results
[2] https://ieeexplore.ieee.org/document/910579
[3] https://nvlabs.github.io/sionna/examples/Bit_Interleaved_Coded_Modulation.html

**Reply by harinduravin** [reply]
@nbecker I think DVB-S2 LDPC encoding is very straight forward, since it is a Repeat Accumulate LDPC code. 

Encoding procedure is given in 

[1] ETSI Standard EN 302 307 V1.4.1. "Digital Video Broadcasting (DVB); Second generation framing structure, channel coding and modulation systems for Broadcasting, Interactive Services, News Gathering and other broadband satellite applications (DVB-S2)." European Telecommunications Standards Institute. Valbonne, France, 2005-03.

If the values from the appendix can be loaded, sparse parity matrix can be formed.

Few lines of Tensorflow or numpy can be used to implement the encoding.

```py
from scipy.sparse import csr_matrix
import tensorflow as tf

parity_matrix = generate_ldpc_H_matrix(n_ldpc=n_ldpc,rate=r,q=18,file_path="ldpccode_short_3_5.txt")
H_csr = csr_matrix(parity_matrix[:, :self.k_ldpc])
self.H_sparse = tf.sparse.SparseTensor(
indices=np.vstack((H_csr.nonzero()[0], H_csr.nonzero()[1])).T.astype(np.int64),
values=H_csr.data.astype(np.int32),  # or int32 if working over integers
dense_shape=H_csr.shape
)

# b_stream is the input message bit tensor
p_bits = tf.transpose(tf.sparse.sparse_dense_matmul(self.H_sparse, tf.transpose(b_stream)))
p_bits = tf.math.mod(p_bits, 2)
p_bits = tf.math.mod(tf.math.cumsum(p_bits, axis=1), 2)
codeword = tf.concat([b_stream, p_bits], axis=1)
```
